File: LeetCode_739/main.cpp
/**
    LeetCode_739
    https://leetcode.com/problems/daily-temperatures/description/
    MatrixDeity, 2018.
*/

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> res(n);
        stack<int> st;
        for (int i = n - 1; i >= 0; --i) {
            while (!st.empty() && temperatures[i] >= temperatures[st.top()])
                st.pop();
            res[i] = !st.empty() ? st.top() - i : 0;
            st.push(i);
        }
        return res;
    }
};

int main() {
    return 0;
}

File: LeetCode_739/main.go
/**
  LeetCode_739
  https://leetcode.com/problems/daily-temperatures/description/
  MatrixDeity, 2022.
*/

package leetcode739

func dailyTemperatures(temperatures []int) []int {
	n := len(temperatures)
	stack := make([]int, 0, n)
	res := make([]int, n)

	for i := n - 1; i >= 0; i-- {
		for len(stack) > 0 && temperatures[stack[len(stack)-1]] <= temperatures[i] {
			stack = stack[:len(stack)-1]
		}

		if len(stack) > 0 {
			res[i] = stack[len(stack)-1] - i
		} else {
			res[i] = 0
		}

		stack = append(stack, i)
	}

	return res
}
